File: PhoneDirectory.py
#All TC on leetcode could not be tested as its premium problem

class PhoneDirectory:

    def __init__(self, maxSlots):
        self.stack = []
        self.s = set()
        for i in range(maxSlots):
            self.stack.append(i)
            self.s.add(i)

    #Here we return the num available from top of the stack
    #Time complexity - O(1)
    #Space complexity - O(1)
    def get(self):
        if len(self.s)<=0:
            return -1
        num = self.stack.pop()
        self.s.remove(num)
        return num

    #Here we check if the num is available in set
    # Time complexity - O(1)
    # Space complexity - O(1)
    def check(self, num):
        return num in self.s

    # Here we check if the num is not available in set and then add to to set and stack
    # Time complexity - O(1)
    # Space complexity - O(1)
    def release(self, num):
        if num not in self.s:
            self.s.add(num)
            self.stack.append(num)


pd = PhoneDirectory(3)
print(pd.get())
print(pd.get())
print(pd.check(0))
print(pd.get())
print(pd.check(0))
print(pd.release(1))
print(pd.check(1))

File: problem1/PhoneDirectory.java
// Time Complexity : Get: O(1), Release: O(1), Check: O(1), n -> Max number
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
package problem1;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

public class PhoneDirectory {
	Set<Integer> numbers;
	Queue<Integer> queue;
	int maxNum;

	/**
	 * Initialize your data structure here
	 * 
	 * @param maxNumbers - The maximum numbers that can be stored in the phone
	 *                   directory.
	 */
	public PhoneDirectory(int maxNumbers) {
		queue = new LinkedList<Integer>();
		for (int i = 0; i < maxNumbers; i++) {
			queue.add(i);
		}
		numbers = new HashSet<Integer>();
		maxNum = maxNumbers;
	}

	/**
	 * Provide a number which is not assigned to anyone.
	 * 
	 * @return - Return an available number. Return -1 if none is available.
	 */
	public int get() {
		if (!queue.isEmpty()) {
			int num = queue.poll();
			numbers.add(num);
			return num;
		}
		return -1;
	}

	/** Check if a number is available or not. */
	public boolean check(int number) {
		if (number < 0 || number >= maxNum) {
			return false;
		}
		return !numbers.contains(number);
	}

	/** Recycle or release a number. */
	public void release(int number) {
		if (numbers.contains(number)) {
			numbers.remove(number);
			queue.add(number);
		}
	}

	public static void main(String[] args) {
		PhoneDirectory obj = new PhoneDirectory(5);
		System.out.println("get(): " + obj.get());
		System.out.println("get(): " + obj.get());
		System.out.println("get(): " + obj.get());
		System.out.println("check(4): " + obj.check(4));
		System.out.println("get(): " + obj.get());
		System.out.println("check(3): " + obj.check(3));
		System.out.println("release(3)");
		obj.release(3);
		System.out.println("check(3): " + obj.check(3));
	}

}
